File: C++/064-最小路径和.cpp
class Solution {
public:
    int minPathSum(vector<vector<int>>& grid) {
        int m = grid.size(); int n = grid[0].size();
        vector<vector<int>> dp(grid);
        for(int i=0;i<m;i++)
            for(int j=0;j<n;j++)
            {
                if( i==0&&j==0 )
                    dp[i][j] = grid[i][j];
                else if( i==0 && j>0 )
                    dp[i][j] = dp[i][j-1]+grid[i][j];
                else if( j==0 && i>0)
                    dp[i][j] = dp[i-1][j]+grid[i][j];
                else
                    dp[i][j] = min( dp[i-1][j] , dp[i][j-1] ) + grid[i][j];
                       
            }
        return dp[m-1][n-1];
    }
};

File: Java/064-最小路径和.java
class Solution {
    public int minPathSum(int[][] grid) {
        int m = grid.length;
        int n = grid[0].length;
        int[][] dp = grid.clone();
        for (int i = 0; i < m; i++)
            for (int j = 0; j < n; j++) {
                if (i == 0 && j == 0)
                    dp[i][j] = grid[i][j];
                else if (i == 0 && j > 0)
                    dp[i][j] = dp[i][j - 1] + grid[i][j];
                else if (j == 0 && i > 0)
                    dp[i][j] = dp[i - 1][j] + grid[i][j];
                else
                    dp[i][j] = Math.min(dp[i - 1][j], dp[i][j - 1]) + grid[i][j];

            }
        return dp[m - 1][n - 1];
    }
}